File: lab5/reduce.js
const reduce = (arr, callback, val) => {
    let res = val;
    for (let i = 0; i < arr.length; i++) {
        res = callback(res, arr[i]);
    }
    return res;
}

const plus = (a, b) => a + b;
const numbers = [1, 2, 3, 4, 5];

console.log(reduce(numbers, plus, 0));
console.log(reduce(numbers, (a, b) => a * b, 1));

File: lab21/reduce.rb
def reduce (arr, val)
    res = val
    for element in arr do
        res = yield(res, element) if block_given?
    end
    p res
end

reduce([1, 2, 3, 4, 5, 6], 0) { |total, v| total + v } # => 21

reduce(["This", "is", "my", "sentence"], "") { |sentence, word| sentence + " " + word } # => "This is my sentence"

reduce([6, 5, 9, 2, 1, 10, 3], Float::INFINITY) do |min, v|
  if min > v
     v
  else
     min
  end
end
# => 1


